adr/0009-extension-structure.adoc

```diff
@@ -0,0 +1,261 @@
+= Extension Structure Guidelines
+
+* Status: Proposed
+* Date: 2025-06-30
+* Authors: @cescoffier, @gsmet
+
+== Context
+
+The Quarkus ecosystem includes more than 700 extensions, many of which were developed without a consistent structure.
+The Quarkus core repository alone contains over 150 extensions.
+
+Most extensions follow the basic runtime/deployment split, but the boundaries between API and internal logic have often blurred.
+Additionally, without clear package structure rules, many split packages have emerged.
+This situation introduces several issues:
+
+* It is unclear which classes are intended as public APIs versus internal implementations.
+* There is no clear boundary between what an extension consumer can depend on and what they should not use.
+* Some extensions mix SPI and implementation code, reducing modularity and increasing the chance of unwanted coupling.
+* The need for build item re-use or coordination between extensions leads to ad-hoc patterns.
+
+More recently, we introduced dev-only modules, further increasing complexity.
+
+To address these issues and prepare for long-term maintainability (and possibly Java Platform Module System (JPMS) adoption), we propose a more structured layout for extension modules.
+
+== Decision
+
+Extensions may adopt a standardized structure, composed of one mandatory and several optional modules, depending on their complexity and integration needs:
+
+* `runtime`: Contains the runtime implementation.
+It may expose public APIs if needed but should prefer delegation to the `runtime-api` module (See <<the-place-of-the-public-api>>).
+This module also contains the `recorders` and other runtime logic.
+* `runtime-api`: Contains public APIs intended to be consumed by other extensions or application code.
+This module should have minimal dependencies.
+Depending on an `runtime-api` module does not transitively include the full extension (i.e., it avoids pulling in runtime or deployment logic). See <<coupling-api-and-spi>> for more details.
+* `deployment`: Contains build-time logic, including `BuildSteps`.
+This module may define internal build items.
+* `deployment-api`: Contains build-time APIs intended to be reused across multiple extensions.
```

Comment:
Can we consider renaming "deployment" to "build"? (same for `deployment-api`)

I understand the module is about "preparing a deployment", but this is super confusing to new contributors/users.

Reply:
Backward compatibility could possible be preserved by falling back to the "deployment" naming, though at a cost. _If_ we care that much about backward compatibility.

Reply:
I really support this idea.